### DISCUSSION TITLE
The ideal container format

### DISCUSSION THREAD

**Discussion by MasterInQuestion** [MasterInQuestion/talk#3] [!Multi-Media]
[[
<b id="issuecomment-1546820018">[ MasterInQuestion @ CE 2024-05-28 18:18:26 UTC:
[ https://github.com/exiftool/exiftool/issues/94#issuecomment-1546820018](https://github.com/exiftool/exiftool/issues/94#issuecomment-1546820018)</b>
    Based on the context of a few posts that have gone missing asserting CFR (Constant Frame Rate) and superiority of MP4.

\----

    MKV may be favored for having less restriction on the media formats (codecs) to contain.
    Any post-analog adequately designed container format (codec too) should be able to handle arbitrary frame rate (and VFR).
    For digital media, there's really no frame rate (nor time base [1]): only presentation timestamps (PTS).

    MP4 is far from the most optimal container format. (so is MKV) [4]
    For the typical application (local media playback), the container may not even be needed.

[ [1]<i id="issuecomment-1546820018-1"></i>
    For typical timestamp implementations: time base shouldn't be of concern.
    But when tailored to optimize for specific applications: time base may actually become a thing.

    Typical high-precision timestamps are based on Int64 and of a granularity of μs (i.e. 1/1,000,000 time base).
    Which should allow a max duration of (( (2^63 - 1) / 1,000,000 / (3,600 * 24) / 365.2425 ≈ 292,277 )) years in this configuration.
    .
    However support for granularity smaller than 1/10,000 may be pointless for common systems. (unreachable target accuracy for underlying system; due to beyond threshold latency [3]) [2]
    And using Int32 instead may provide relatively significant performance benefit in certain cases.
    .
    Thus an alternative configuration based on Int32 and of granularity of ms (1/1,000 s) is plausible:
    Which should still allow a max duration of (( (2^31 - 1) / 1,000 / (3,600 * 24) ≈ 24.8551348 )) days. (and max 1,000 FPS) ]

[ [2]<i id="issuecomment-1546820018-2"></i>
    Significant amount of audio handling are of 48,000 Hz. (1/48,000 < 1/10,000 indeed)
    But what handles the final audio playback is typically dedicated hardware (not software based on CPU [ Precisely, general purpose processors. ]).
    And which should load far more than 1 sample at once (more than 1 sample per frame), and is [FIFO](https://en.wikipedia.org/wiki/FIFO_(computing_and_electronics)) like.

    Note audio playback is somewhat different than video: the sample value (content) is uttered at the timestamp but not maintained across the whole duration. ]

[ [3]<i id="issuecomment-1546820018-3"></i>
    The system's latency may be estimated using &#x60;while ( 1 ) { printTS(); };&#x60; alike structure.
    E.g. in Bash:
[[
```bash
	x0="${EPOCHREALTIME%%.*}";
	while true;
	do {
	x1="$EPOCHREALTIME";
	x2="${x1%%.*}";
	[ "$x2" == "$x0" ] && continue;
	echo -E "$x1";
	x0="$x2";
	};
	done;
#
#	Less likely to work:
# [[
#	while true; do { [ "${EPOCHREALTIME##*.}" -eq 0 ] && echo -E "$EPOCHREALTIME"; }; done;
# ]]
#
```
]] ]

[ [4]<i id="issuecomment-1546820018-4"></i>
    Why exactly?
    MP4 had the information critical [1] for media playback stored at 1 location (usually near the end of file; or the start of file).
    Such design has caused some problems. [ See [FFmpeg's documentation](https://ffmpeg.org/ffmpeg-formats.html#matroska) on related muxers. ]
    .
    Timestamps could be wrapped as part of each packet (encoded frame), thereby achieve superior performance meanwhile getting rid of all aforementioned problems.
    Such was somewhat done in MKV. But of significant overhead. (complicated "cues", and many more)

    More on the topic:
    [ https://github.com/MasterInQuestion/talk/discussions/3](https://github.com/MasterInQuestion/talk/discussions/3)

[1] "moov atom", in MP4 terminology. Mostly contained the presentation timestamps, and frame dimensions. ]

    ----

    The ideal container format should be based on a general purpose container format: 7z should be a good start point. <b>]</b>

\----

<b id="issuecomment-1546859236">[ MasterInQuestion @ CE 2023-05-14 09:55:21 UTC:
[ https://github.com/exiftool/exiftool/issues/94#issuecomment-1546859236](https://github.com/exiftool/exiftool/issues/94#issuecomment-1546859236)</b>
    [https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers#mpeg-4_mp4](https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers#mpeg-4_mp4)

    The description covers more than just H.264: intended for all codecs (both audio/video) in general.
    It's a functional analysis on the underpinnings: whose validity is irrelevant of whether particular implementations work or not.

    Unnecessary elements may indeed exist (or even be required).
    But indeed they do not have to. <b>]</b>
]]

**Comment by MasterInQuestion** [comment]
Name it MPEG-X, essentially 7z archive with alike content:
[ "eg.mpx"
    "etc/Comment on Random Non-Sense.txt"
    "!.mpp" [ Media Presentation Profile ]
    "a1.flac"
    "a2.opus"
    "a3.vorbis"
    "a4.aac"
    "v1.vp9"
    "v2.h264"
    "v3.h265"
    "v4.av1" ]

    It could also sort of address some legacy JPEG problems:
[ "eg.jpx"
    "!.mpp" [ Color Profile etc. ]
    "v.jpg" ]
    Some JPEG may receive further size reduction this way... Real losslessly.
[ ^
    Probably meaningless:
    [ https://github.com/libjxl/libjxl/issues/432#issuecomment-2465865410](https://github.com/libjxl/libjxl/issues/432#issuecomment-2465865410)
    ; if the re-encoding wouldn't introduce unacceptable quality loss. ]
\
\
    Media Presentation Profile is meant to describe whatsoever possible rendering in Plain Text.
    Complex blob like [MXF](https://www.google.com/search?hl=en&gl=ca&num=10&q=MXF), [PSD](https://en.wikipedia.org/wiki/Adobe_Photoshop#File_format), [3DS](https://en.wikipedia.org/wiki/.3ds), [DWG](https://en.wikipedia.org/wiki/AutoCAD#File_formats) etc. [1] would there be pointless.
    [ This should also supersede MIDI audio. Extending further could even become a game engine..? (and somewhat complement HTML) ]

    For some cases a single ".mpp" file should suffice. (XVG)

    Attention shall be taken to not retrace the serious failure of SVG:
    [ https://github.com/MasterInQuestion/Markup/issues/2](https://github.com/MasterInQuestion/Markup/issues/2)
    https://github.com/MasterInQuestion/Markup/blob/main/Junkyard%20Tornado.svg?short_path=1

[1] https://www.google.com/search?hl=en&gl=us&num=100&q=site:adobe.com/creativecloud/file-types+PSD|3DS|DWG
\
\
    The Presentation Profile shall be more focused on the Presentation part:
    I.e. there shall be the distinction between "sources" and "post-rendering" (manipulation based on the sources).

    Which means the MPP itself doesn't promote (though support) embedding raw vector sources right inline.
    (using independent files and referencing shall be preferable)

    The format defining such vector sources shall be named XVG.
    Note the Graphics may contain audio too.
    .
    There shall be no essential difference between MPP/XVG:
    XVG is essentially MPP without referring any external source. (thus independently presentable)

**Comment by MasterInQuestion** [comment]
The simplest exemplar:
    "_00_00_00.000000.webp"
    "_00_00_00.033333.webp"
    "_00_00_00.066667.webp"
    "_00_00_00.100000.webp"
    "_10_00_00.000000.jxl"
    "_10_00_01.000000" -> "./1__00_00_00.000000.jxl" [ Symlink. ]
    "_10_00_01.500000" -> "./_00_00_00.066667.webp"
    "_10_00_02.000000.webp" -> "./_00_00_00.100000.webp"
    "_23_59_59.999999.heif"
    "1__00_00_00.000000.jxl"

    Done with the SMPTE non-sense:
    https://trac.ffmpeg.org/ticket/11249#comment:13

\
    MPX Simple:
    "_00_00_01.000.vpx"
    "_00_00_02.000" -> "./_00_00_01.000.vpx#2"
    "_00_00_02.125" -> "./_00_00_01.000.vpx#3"
    "_00_00_02.500" -> "./_00_00_01.000.vpx#9223372036854775807"
    "_00_00_03.000" -> "./_00_00_01.000.vpx#4"
    ...
    "106751991__00_00_00.000.vpx"
.
    Support notation for Inter-frame compression.

    /<a href="https://regex101.com/?regex=%23%5Cd%2B%24&flags">`#\d+$`</a>/ links are technically broken symlinks:
    But special-handled by the media parser.

    VPX is a future video/image codec, that handles GOP (Group of Pictures).
    See also: https://trac.ffmpeg.org/ticket/11563#comment:5

    The 7z archive wrapper in certain cases maybe actually skipped: without practical difference.
    (name the container directory "eg.mpx" etc. could suffice)

**Comment by MasterInQuestion** [comment]
= Where major current container formats exactly failed =

    ...

**Comment by beedell-roke** [comment]
> https://github.com/MasterInQuestion/talk/discussions/3#discussion-5219034

@MasterInQuestion, the formatting in this thread is utterly unreadable. What's happened?

**Reply by MasterInQuestion** [reply]
Seemingly alright for me? (normal web interface)
    How did you view it?

    Some portions may require adequate technical background to understand.
    Google Search the keywords should work.

    ----

    For the linked post: https://trac.ffmpeg.org/ticket/4907#comment:33
    The key idea is:
    There shall be no unnecessary wrapping that pointlessly sophisticates the implementation.
    (alike animation bundles are less interoperable than regular files)

**Reply by beedell-roke** [reply]
> https://github.com/MasterInQuestion/talk/discussions/3#discussioncomment-11521783

@MasterInQuestion, I'm not referring to terminology. I'm referring to markup syntax:

1. There are 5 spaces (`     `) between each sentence.
2. The start of each hyperlink is formatted in an invalid manner (`[ placeholder` instead of `[placeholder`) which means that it doesn't render as an `<a href=""></a>`.
3. What seem to be list entries aren't actually `<li>`s, because they're delineated by `[ [1]` instead of `1. `.

Perhaps converting it into its source demonstrates the errata better:

<details>

> ~~~Markdown
> [[ [ MasterInQuestion @ CE 2024-05-28 18:18:26 UTC: [ https://github.com/exiftool/exiftool/issues/94#issuecomment-1546820018](https://github.com/exiftool/exiftool/issues/94#issuecomment-1546820018)     Based on the context of a few posts that have gone missing asserting CFR (Constant Frame Rate) and superiority of MP4.
> 
> ----
> 
>     MKV may be favored for having less restriction on the media formats (codecs) to contain.     Any post-analog adequately designed container format (codec too) should be able to handle arbitrary frame rate (and VFR).     For digital media, there's really no frame rate (nor time base [1]): only presentation timestamps (PTS).
> 
>     MP4 is far from the most optimal container format. (so is MKV) [4]     For the typical application (local media playback), the container may not even be needed.
> 
> [ [1]     For typical timestamp implementations: time base shouldn't be of concern.     But when tailored to optimize for specific applications: time base may actually become a thing.
> 
>     Typical high-precision timestamps are based on Int64 and of a granularity of μs (i.e. 1/1,000,000 time base).     Which should allow a max duration of (( (2^63 - 1) / 1,000,000 / (3,600 * 24) / 365.2425 ≈ 292,277 )) years in this configuration.     .     However support for granularity smaller than 1/10,000 may be pointless for common systems. (unreachable target accuracy for underlying system; due to beyond threshold latency [3]) [2]     And using Int32 instead may provide relatively significant performance benefit in certain cases.     .     Thus an alternative configuration based on Int32 and of granularity of ms (1/1,000 s) is plausible:     Which should still allow a max duration of (( (2^31 - 1) / 1,000 / (3,600 * 24) ≈ 24.8551348 )) days. (and max 1,000 FPS) ]
> 
> [ [2]     Significant amount of audio handling are of 48,000 Hz. (1/48,000 < 1/10,000 indeed)     But what handles the final audio playback is typically dedicated hardware (not software based on CPU [ Precisely, general purpose processors. ]).     And which should load far more than 1 sample at once (more than 1 sample per frame), and is [FIFO](https://en.wikipedia.org/wiki/FIFO_(computing_and_electronics)) like.
> 
>     Note audio playback is somewhat different than video: the sample value (content) is uttered at the timestamp but not maintained across the whole duration. ]
> 
> [ [3]     The system's latency may be estimated using `while ( 1 ) { printTS(); };` alike structure.     E.g. in Bash: [[
> 
> ```shell
> 	x0="${EPOCHREALTIME%%.*}";
> 	while true;
> 	do {
> 	x1="$EPOCHREALTIME";
> 	x2="${x1%%.*}";
> 	[ "$x2" == "$x0" ] && continue;
> 	echo -E "$x1";
> 	x0="$x2";
> 	};
> 	done;
> #
> #	Less likely to work:
> # [[
> #	while true; do { [ "${EPOCHREALTIME##*.}" -eq 0 ] && echo -E "$EPOCHREALTIME"; }; done;
> # ]]
> #
> ```
> 
> ]] ]
> 
> [ [4]     Why exactly?     MP4 had the information critical [1] for media playback stored at 1 location (usually near the end of file; or the start of file).     Such design has caused some problems. [ See [FFmpeg's documentation](https://ffmpeg.org/ffmpeg-formats.html#matroska) on related muxers. ]     .     Timestamps could be wrapped as part of each packet (encoded frame), thereby achieve superior performance meanwhile getting rid of all aforementioned problems.     Such was somewhat done in MKV. But of significant overhead. (complicated "cues", and many more)
> 
>     More on the topic:     [ https://github.com/MasterInQuestion/talk/discussions/3](https://github.com/MasterInQuestion/talk/discussions/3)
> 
> [1] "moov atom", in MP4 terminology. Mostly contained the presentation timestamps, and frame dimensions. ]
> 
>     ----
> 
>     The ideal container format should be based on a general purpose container format: 7z should be a good start point. ]
> 
> ----
> 
> [ MasterInQuestion @ CE 2023-05-14 09:55:21 UTC: [ https://github.com/exiftool/exiftool/issues/94#issuecomment-1546859236](https://github.com/exiftool/exiftool/issues/94#issuecomment-1546859236)     https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers#mpeg-4_mp4
> 
>     The description covers more than just H.264: intended for all codecs (both audio/video) in general.     It's a functional analysis on the underpinnings: whose validity is irrelevant of whether particular implementations work or not.
> 
>     Unnecessary elements may indeed exist (or even be required).     But indeed they do not have to. ] ]]
>
> ~~~

</details>

**Reply by MasterInQuestion** [reply]
For Markup/Markdown related:
    [ https://github.com/MasterInQuestion/talk/discussions/28](https://github.com/MasterInQuestion/talk/discussions/28)

    And I'm avoiding the over-interpretation of Markdown.
    (everything shall be Plain Text identical, ideally)

    I do exhaustive research on all possibilities of valid representations.
    And the current scheme is the most viable among all others.
    (best readability possible; and general interoperability)
\
\
    Indentation is important to differentiate auto-wrapped lines from adjacent next line.
    And the "`[${Tag}]`" marks function more like footnotes.

    Does the formatting somehow appear unintended in some interfaces?
    E.g. GitHub apps (Desktop/Mobile), email notifications
    ; which I don't use and haven't verified.

**Reply by beedell-roke** [reply]
> (everything shall be Plain Text identical, ideally)

@MasterInQuestion, that's not possible in GitHub-flavoured Markdown, unfortunately. Both you and I would agree that plain text involves indentation to maintain readability, I presume? That's impossible in GHFM, because the renderer doesn't support anything like https://github.com/searKing/preview-vscode/issues/96#issuecomment-2504553232. Even then, it shan't be rendered.

> Indentation is important to differentiate auto-wrapped lines from adjacent next line.

That's *very* true, except that it's meant to be achieved via CSS. Wrangling HTML (generated from Markdown) that isn't styled in that manner isn't the way to achieve it. It breaks all other features that you've attempted to employ, as verbosely aforestated.

> And the "`[${Tag}]`" marks function more like footnotes.

They would, if you were using real footnote syntax. You're not - you're using link abstraction syntax. What you're looking for is `[^1]` for references, and `[^1]: https://example.com/` for bibliographic assignments.

Don't just invent syntaxes - I've created conlangs, and it's a multi-year undertaking that usually renders itself fruitless.

I'll provide an example of what actually *is* valid MD:
<details>

~~~Markdown
MasterInQuestion @ CE 2024-05-28 18:18:26 UTC: [https://github.com/exiftool/exiftool/issues/94#issuecomment-1546820018](https://github.com/exiftool/exiftool/issues/94#issuecomment-1546820018) Based on the context of a few posts that have gone missing asserting CFR (Constant Frame Rate) and superiority of MP4.

----

MKV may be favored for having less restriction on the media formats (codecs) to contain. Any post-analog adequately designed container format (codec too) should be able to handle arbitrary frame rate (and VFR). For digital media, there's really no frame rate (nor time base [^1]): only presentation timestamps (PTS).

MP4 is far from the most optimal container format. (so is MKV) [^4] For the typical application (local media playback), the container may not even be needed.

[^1]: For typical timestamp implementations: time base shouldn't be of concern. But when tailored to optimize for specific applications: time base may actually become a thing.

      Typical high-precision timestamps are based on Int64 and of a granularity of μs (i.e. 1/1,000,000 time base). Which should allow a max duration of (( (2^63 - 1) / 1,000,000 / (3,600 * 24) / 365.2425 ≈ 292,277 )) years in this configuration. However support for granularity smaller than 1/10,000 may be pointless for common systems. (unreachable target accuracy for underlying system; due to beyond threshold latency [^3]) [^2] And using Int32 instead may provide relatively significant performance benefit in certain cases. Thus an alternative configuration based on Int32 and of granularity of ms (1/1,000 s) is plausible: Which should still allow a max duration of (( (2^31 - 1) / 1,000 / (3,600 * 24) ≈ 24.8551348 )) days. (and max 1,000 FPS)

[^2]: Significant amount of audio handling are of 48,000 Hz. (1/48,000 < 1/10,000 indeed) But what handles the final audio playback is typically dedicated hardware (not software based on CPU [Precisely, general purpose processors.]). And which should load far more than 1 sample at once (more than 1 sample per frame), and is [FIFO](https://en.wikipedia.org/wiki/FIFO_(computing_and_electronics)) like.

      Note audio playback is somewhat different than video: the sample value (content) is uttered at the timestamp but not maintained across the whole duration.

[^3] The system's latency may be estimated using `while ( 1 ) { printTS(); };` alike structure. E.g. in Bash:

     ```shell
     x0="${EPOCHREALTIME%%.*}";
     while true;
     do {
     x1="$EPOCHREALTIME";
     x2="${x1%%.*}";
     ["$x2" == "$x0"] && continue;
     echo -E "$x1";
     x0="$x2";
     };
     done;
     #
     #	Less likely to work:
     # [[
     #	while true; do { ["${EPOCHREALTIME##*.}" -eq 0] && echo -E "$EPOCHREALTIME"; }; done;
     #]]
     #
     ```

[^4]: Why exactly? MP4 had the information critical [^1] for media playback stored at 1 location (usually near the end of file; or the start of file). Such design has caused some problems. [See [FFmpeg's documentation](https://ffmpeg.org/ffmpeg-formats.html#matroska) on related muxers.] . Timestamps could be wrapped as part of each packet (encoded frame), thereby achieve superior performance meanwhile getting rid of all aforementioned problems. Such was somewhat done in MKV. But of significant overhead. (complicated "cues", and many more)

More on the topic: [https://github.com/MasterInQuestion/talk/discussions/3](https://github.com/MasterInQuestion/talk/discussions/3)

1. "moov atom", in MP4 terminology. Mostly contained the presentation timestamps, and frame dimensions.

---

The ideal container format should be based on a general purpose container format: 7z should be a good start point.

---

MasterInQuestion @ CE 2023-05-14 09:55:21 UTC: [https://github.com/exiftool/exiftool/issues/94#issuecomment-1546859236](https://github.com/exiftool/exiftool/issues/94#issuecomment-1546859236) https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers#mpeg-4_mp4

The description covers more than just H.264: intended for all codecs (both audio/video) in general. It's a functional analysis on the underpinnings: whose validity is irrelevant of whether particular implementations work or not.

Unnecessary elements may indeed exist (or even be required). But indeed they do not have to.

~~~
</details>

I've never seen self-referencing footnotes until now, but it *is* valid.

**Reply by MasterInQuestion** [reply]
Thanks for the formatting insights.
    Probably you missed 1 significant point:
    "What's the purpose of all these sophisticated markups if things may be very well clearly expressed in Plain Text?"

    My primary focus is to avoid the over-interpretation had rendered texts, unjustifiably differed from their Plain Text source.
    The Markdown footnotes create much unintentional changes.

    I take every step cautiously.
    Before coining things I do analyze all other practical solutions.

    Compare the rendered side-by-side.
    The distinction shall be apparent.

**Reply by beedell-roke** [reply]
> https://github.com/MasterInQuestion/talk/discussions/3#discussioncomment-11524968

@MasterInQuestion, you're missing the point. As it is, your markup is rendered as Markdown on GitHub. There is no option to choose the IANA Media Type you want your text rendered as, outside of files in repositories.

This means that you not adhering to GH's MD implementation has rendered your comment nigh unreadable. The undermentioned is what I see:
https://github.com/MasterInQuestion/attach/raw/main/MasterInQuestion/talk/3/texts-rendered.webp

**Reply by MasterInQuestion** [reply]
The content appears within the expectation.

    The syntax of this document is sort of an experiment:
    Demonstrating the referencing structures alike in programming languages.
    (lookup and scopes)

    It's possible to be written in some other less "GOTO", but more verbose manner.
    Which is somewhat not of my liking.
\
\
    Actually the entire main post is a thread gather dump:
    Sort of database schema implemented in Plain Text.

    I guess your parsing difficulties primarily originated from not recognizing the outline.
    (thus failed to separate the contexts)
    Some simple standard bracket matching algorithm should help.
    Alike: https://github.com/MasterInQuestion/attach/raw/main/webmproject/libwebp/33357131698ba0abb5b7a32b6f274f6f58170642/Inflammatory.40.re.webp

**Comment by MasterInQuestion** [comment]
<b id="issuecomment-2910107688">[ mckaygerhard @ CE 2025-05-26 15:31:33 UTC:
[ https://github.com/exiftool/exiftool/issues/94#issuecomment-2910107688](https://github.com/exiftool/exiftool/issues/94#issuecomment-2910107688)</b>
    "Based on the context of a few posts that have gone missing asserting CFR (Constant Frame Rate) and superiority of MP4."

    Hi @MasterInQuestion and all here!
    It would be wonderful to have support for writing in Matroska formats, but...

    But Matroska, this format is very cumbersome... as yourself point out in the cited content:
    MP4 is much more optimal, even though it's not much better quality as noted in comparison with MP4.

    As demonstration why MP4 is "more fit and best for":
    My older devices can play MP4 effortlessly, while Matroska with the same configuration take much longer and require much more power...
    Add this to the battery consumption on mobile devices if you play Matroska files.

    Note:
    Yes, MP4 design has caused some problems. [ See [FFmpeg's documentation](https://ffmpeg.org/ffmpeg-formats.html#matroska) on related muxers. ]
    By example incongruent timed text for subtitles...
    But is far more light to play and less energy consumption. <b>]</b>
.
    @mckaygerhard
    I fear you missed much of the points...
    (and likely there're significant errors in your statements)

    Do note that:
    The container is mere container: it doesn't affect the media content (codec).
    That causes alike notable differences, could only because of implementation glitches.
    (core logic between the 2 containers shouldn't differ much)

    Your observed difference is typically resulted from different media codecs used. (container irrelevant)
